File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Shared/PitchGraphView.swift
import SwiftUI
import Charts
import VoxaTrace

/// A pitch point for charting purposes.
struct PitchDataPoint: Identifiable {
    let id = UUID()
    let time: Float
    let pitch: Float  // In MIDI note numbers (e.g., 60 = C4)
    let series: String      // User-facing series name (for legend)
    let segmentId: String   // Internal segment ID (series + segment number)
}

/// Segments a pitch contour into consecutive voiced regions.
/// Returns array of PitchDataPoint with MIDI note numbers.
/// - Parameter inputIsMidi: If true, pitches are already in MIDI format (skip Hz→MIDI conversion)
private func segmentPitchContour(
    times: [Float],
    pitches: [Float],
    seriesName: String,
    inputIsMidi: Bool = false,
    gapThreshold: Float = 0.025  // 25ms gap threshold (2.5x the 10ms hop)
) -> [PitchDataPoint] {
    var points: [PitchDataPoint] = []
    var segmentIndex = 0
    var lastValidTime: Float? = nil

    for (time, pitch) in zip(times, pitches) {
        let midiNote: Float
        if inputIsMidi {
            // Input is already MIDI - use directly
            midiNote = pitch
        } else {
            // Input is Hz - convert to MIDI
            midiNote = CalibraMusic.hzToMidi(pitch)
        }

        // Skip invalid pitches (unvoiced sentinel is -1 for MIDI, <= 0 for Hz)
        if midiNote.isNaN || midiNote <= 0 {
            // Gap detected - next valid point starts a new segment
            lastValidTime = nil
            continue
        }

        // Check for time gap (unvoiced region passed)
        if let last = lastValidTime, (time - last) > gapThreshold {
            segmentIndex += 1
        }

        let segmentId = "\(seriesName)_\(segmentIndex)"
        points.append(PitchDataPoint(
            time: time,
            pitch: midiNote,
            series: seriesName,
            segmentId: segmentId
        ))
        lastValidTime = time
    }

    return points
}

/// A reusable pitch contour graph using Swift Charts.
///
/// Displays pitch values over time with musical note labels on Y-axis.
/// Supports both single contour and multi-contour overlay modes.
struct PitchGraphView: View {
    let dataPoints: [PitchDataPoint]
    let title: String?
    let height: CGFloat
    let seriesColors: [String: Color]
    let seriesOrder: [String]

    init(
        pitchesHz: [Float],
        times: [Float]? = nil,
        color: Color = .blue,
        title: String? = nil,
        series: String = "Pitch",
        height: CGFloat = 200,
        inputIsMidi: Bool = false
    ) {
        // Generate time values if not provided (assuming 10ms hop)
        let timeValues = times ?? pitchesHz.indices.map { Float($0) * 0.01 }

        // Segment the contour to avoid connecting across gaps
        self.dataPoints = segmentPitchContour(
            times: timeValues,
            pitches: pitchesHz,
            seriesName: series,
            inputIsMidi: inputIsMidi
        )
        self.title = title
        self.height = height
        self.seriesColors = [series: color]
        self.seriesOrder = [series]
    }

    /// Initialize with multiple pitch contours for overlay comparison.
    /// This version uses a shared times array for all contours (legacy mode).
    init(
        contours: [(pitches: [Float], color: Color, label: String)],
        times: [Float]? = nil,
        title: String? = nil,
        height: CGFloat = 200,
        inputIsMidi: Bool = false
    ) {
        var allPoints: [PitchDataPoint] = []
        var colors: [String: Color] = [:]
        var order: [String] = []

        for contour in contours {
            let timeValues = times ?? contour.pitches.indices.map { Float($0) * 0.01 }
            // Segment each contour to avoid connecting across gaps
            let points = segmentPitchContour(
                times: timeValues,
                pitches: contour.pitches,
                seriesName: contour.label,
                inputIsMidi: inputIsMidi
            )
            allPoints.append(contentsOf: points)
            colors[contour.label] = contour.color
            order.append(contour.label)
        }

        self.dataPoints = allPoints
        self.title = title
        self.height = height
        self.seriesColors = colors
        self.seriesOrder = order
    }

    /// Initialize with multiple pitch contours, each with its own times array.
    /// This ensures proper time alignment when contours have different frame counts.
    init(
        contoursWithTimes: [(times: [Float], pitches: [Float], color: Color, label: String)],
        title: String? = nil,
        height: CGFloat = 200,
        inputIsMidi: Bool = false
    ) {
        var allPoints: [PitchDataPoint] = []
        var colors: [String: Color] = [:]
        var order: [String] = []

        for contour in contoursWithTimes {
            // Each contour uses its own times array
            let points = segmentPitchContour(
                times: contour.times,
                pitches: contour.pitches,
                seriesName: contour.label,
                inputIsMidi: inputIsMidi
            )
            allPoints.append(contentsOf: points)
            colors[contour.label] = contour.color
            order.append(contour.label)
        }

        self.dataPoints = allPoints
        self.title = title
        self.height = height
        self.seriesColors = colors
        self.seriesOrder = order
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let title = title {
                Text(title)
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            Chart(dataPoints) { point in
                LineMark(
                    x: .value("Time (s)", point.time),
                    y: .value("MIDI", point.pitch),
                    series: .value("Segment", point.segmentId)
                )
                // Color by series name (not segment) so all segments of same series have same color
                .foregroundStyle(by: .value("Series", point.series))
                .lineStyle(StrokeStyle(lineWidth: 2))
            }
            // Map series names to colors
            .chartForegroundStyleScale(domain: seriesOrder, range: seriesOrder.map { seriesColors[$0] ?? .gray })
            .chartYScale(domain: yAxisDomain)
            .chartXAxisLabel("Time (s)")
            .chartYAxisLabel("Note")
            .chartYAxis {
                AxisMarks(values: yAxisTickValues) { value in
                    AxisGridLine()
                    AxisTick()
                    AxisValueLabel {
                        if let midi = value.as(Float.self) {
                            Text(CalibraMusic.midiToNoteLabel(midi))
                                .font(.caption2)
                        }
                    }
                }
            }
            .chartLegend(seriesNames.count > 1 ? .visible : .hidden)
            .frame(height: height)
        }
        .padding(8)
        .background(Color(.secondarySystemBackground))
        .cornerRadius(8)
    }

    /// Calculate Y-axis domain based on MIDI note range.
    private var yAxisDomain: ClosedRange<Float> {
        let pitches = dataPoints.map { $0.pitch }
        guard !pitches.isEmpty else {
            // Default to C3-C5 range (MIDI 48-72)
            return 48...72
        }

        let minMidi = pitches.min() ?? 48
        let maxMidi = pitches.max() ?? 72

        // Round to nearest semitone and add padding (at least 2 semitones each side)
        var lowerBound = floor(minMidi) - 2
        var upperBound = ceil(maxMidi) + 2

        // Enforce minimum range of 12 semitones (1 octave) for visual clarity
        let minRange: Float = 12.0
        let currentRange = upperBound - lowerBound
        if currentRange < minRange {
            let expansion = (minRange - currentRange) / 2.0
            lowerBound -= expansion
            upperBound += expansion
        }

        return lowerBound...upperBound
    }

    /// Generate tick values at whole semitone boundaries within the domain.
    private var yAxisTickValues: [Float] {
        let domain = yAxisDomain
        let range = domain.upperBound - domain.lowerBound

        // Determine stride: every semitone if range <= 12, otherwise thin out
        let stride: Float
        if range <= 12 {
            stride = 1.0  // Every semitone
        } else if range <= 24 {
            stride = 2.0  // Every whole tone
        } else {
            stride = 3.0  // Every minor third
        }

        var ticks: [Float] = []
        var current = ceil(domain.lowerBound)
        while current <= domain.upperBound {
            ticks.append(current)
            current += stride
        }

        return ticks
    }

    /// Get unique series names for legend.
    private var seriesNames: Set<String> {
        Set(dataPoints.map { $0.series })
    }
}

/// A specialized overlay graph for comparing raw vs processed pitch contours.
struct PitchComparisonGraphView: View {
    let rawPitches: [Float]
    let processedPitches: [Float]
    let times: [Float]?
    let rawLabel: String
    let processedLabel: String
    let height: CGFloat

    init(
        rawPitches: [Float],
        processedPitches: [Float],
        times: [Float]? = nil,
        rawLabel: String = "Raw",
        processedLabel: String = "Processed",
        height: CGFloat = 200
    ) {
        self.rawPitches = rawPitches
        self.processedPitches = processedPitches
        self.times = times
        self.rawLabel = rawLabel
        self.processedLabel = processedLabel
        self.height = height
    }

    var body: some View {
        PitchGraphView(
            contours: [
                (pitches: rawPitches, color: .gray, label: rawLabel),
                (pitches: processedPitches, color: .blue, label: processedLabel)
            ],
            times: times,
            title: "Raw vs Processed",
            height: height
        )
    }
}

/// Multi-contour overlay for post-processing demo.
struct MultiContourGraphView: View {
    struct ContourData {
        let pitches: [Float]
        let label: String
        let color: Color
        let isVisible: Bool
    }

    let contours: [ContourData]
    let times: [Float]?
    let height: CGFloat

    var body: some View {
        let visibleContours = contours.filter { $0.isVisible }

        VStack(alignment: .leading, spacing: 4) {
            Text("Pitch Contours")
                .font(.caption)
                .foregroundColor(.secondary)

            if visibleContours.isEmpty {
                Text("Select contours to display")
                    .foregroundColor(.secondary)
                    .frame(height: height)
                    .frame(maxWidth: .infinity)
            } else {
                PitchGraphView(
                    contours: visibleContours.map { ($0.pitches, $0.color, $0.label) },
                    times: times,
                    height: height
                )
            }
        }
    }
}

// MARK: - Preview

#Preview {
    VStack(spacing: 16) {
        // Single contour - G3 to E4 range
        PitchGraphView(
            pitchesHz: [196, 220, 247, -1, -1, 262, 277, 294, 311, 330],
            title: "Single Contour (G3-E4)"
        )

        // Comparison - showing raw (noisy) vs processed (smooth)
        PitchComparisonGraphView(
            rawPitches: [196, 392, 220, 208, 440, 247, 262, 277],
            processedPitches: [196, 196, 208, 208, 220, 247, 262, 277]
        )
    }
    .padding()
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/shared/PitchGraphView.kt
package com.musicmuni.voxatrace.demo.sections.shared

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.musicmuni.voxatrace.calibra.CalibraMusic
import kotlin.math.ceil
import kotlin.math.floor

/**
 * Data point for a pitch contour segment.
 */
data class PitchDataPoint(
    val time: Float,
    val pitch: Float,  // In MIDI note numbers
    val series: String,
    val segmentId: String
)

/**
 * Segments a pitch contour into consecutive voiced regions.
 * Returns array of PitchDataPoint with MIDI note numbers.
 *
 * @param inputIsMidi If true, pitches are already in MIDI format (skip Hz->MIDI conversion)
 * @param gapThreshold Time gap in seconds to start a new segment (default 25ms)
 */
private fun segmentPitchContour(
    times: List<Float>,
    pitches: List<Float>,
    seriesName: String,
    inputIsMidi: Boolean = false,
    gapThreshold: Float = 0.025f
): List<PitchDataPoint> {
    val points = mutableListOf<PitchDataPoint>()
    var segmentIndex = 0
    var lastValidTime: Float? = null

    for ((time, pitch) in times.zip(pitches)) {
        val midiNote: Float = if (inputIsMidi) {
            pitch
        } else {
            CalibraMusic.hzToMidi(pitch)
        }

        // Skip invalid pitches (unvoiced sentinel is -1 for MIDI, <= 0 for Hz)
        if (midiNote.isNaN() || midiNote <= 0) {
            lastValidTime = null
            continue
        }

        // Check for time gap (unvoiced region passed)
        lastValidTime?.let { last ->
            if ((time - last) > gapThreshold) {
                segmentIndex++
            }
        }

        val segmentId = "${seriesName}_$segmentIndex"
        points.add(PitchDataPoint(
            time = time,
            pitch = midiNote,
            series = seriesName,
            segmentId = segmentId
        ))
        lastValidTime = time
    }

    return points
}

/**
 * Contour data for multi-contour graphs.
 */
data class ContourData(
    val pitches: List<Float>,
    val color: Color,
    val label: String,
    val times: List<Float>? = null
)

/**
 * A reusable pitch contour graph using Compose Canvas.
 *
 * Displays pitch values over time with musical note labels on Y-axis.
 * Supports both single contour and multi-contour overlay modes.
 */
@Composable
fun PitchGraphView(
    pitchesHz: List<Float>,
    times: List<Float>? = null,
    color: Color = Color.Blue,
    title: String? = null,
    series: String = "Pitch",
    height: Int = 200,
    inputIsMidi: Boolean = false
) {
    val timeValues = times ?: pitchesHz.indices.map { it * 0.01f }
    val dataPoints = remember(pitchesHz, times, series, inputIsMidi) {
        segmentPitchContour(timeValues, pitchesHz, series, inputIsMidi)
    }
    val seriesColors = mapOf(series to color)

    PitchGraphContent(
        dataPoints = dataPoints,
        seriesColors = seriesColors,
        title = title,
        height = height
    )
}

/**
 * Initialize with multiple pitch contours for overlay comparison.
 * Uses a shared times array for all contours.
 */
@Composable
fun PitchGraphView(
    contours: List<ContourData>,
    times: List<Float>? = null,
    title: String? = null,
    height: Int = 200,
    inputIsMidi: Boolean = false
) {
    val dataPoints = remember(contours, times, inputIsMidi) {
        contours.flatMap { contour ->
            val timeValues = contour.times ?: times ?: contour.pitches.indices.map { it * 0.01f }
            segmentPitchContour(timeValues, contour.pitches, contour.label, inputIsMidi)
        }
    }
    val seriesColors = contours.associate { it.label to it.color }

    PitchGraphContent(
        dataPoints = dataPoints,
        seriesColors = seriesColors,
        title = title,
        height = height
    )
}

/**
 * Initialize with multiple pitch contours, each with its own times array.
 */
@Composable
fun PitchGraphViewWithTimes(
    contoursWithTimes: List<Triple<List<Float>, List<Float>, Pair<Color, String>>>,
    title: String? = null,
    height: Int = 200,
    inputIsMidi: Boolean = false
) {
    val dataPoints = remember(contoursWithTimes, inputIsMidi) {
        contoursWithTimes.flatMap { (times, pitches, colorLabel) ->
            segmentPitchContour(times, pitches, colorLabel.second, inputIsMidi)
        }
    }
    val seriesColors = contoursWithTimes.associate { it.third.second to it.third.first }

    PitchGraphContent(
        dataPoints = dataPoints,
        seriesColors = seriesColors,
        title = title,
        height = height
    )
}

@Composable
private fun PitchGraphContent(
    dataPoints: List<PitchDataPoint>,
    seriesColors: Map<String, Color>,
    title: String?,
    height: Int
) {
    val surfaceColor = MaterialTheme.colorScheme.surfaceVariant
    val onSurfaceColor = MaterialTheme.colorScheme.onSurfaceVariant
    val gridColor = MaterialTheme.colorScheme.outline.copy(alpha = 0.3f)

    // Calculate domain
    val (yMin, yMax, timeMin, timeMax) = remember(dataPoints) {
        if (dataPoints.isEmpty()) {
            listOf(48f, 72f, 0f, 1f)
        } else {
            val pitches = dataPoints.map { it.pitch }
            val times = dataPoints.map { it.time }
            val minMidi = floor(pitches.min()) - 2
            val maxMidi = ceil(pitches.max()) + 2
            listOf(minMidi, maxMidi, times.min(), times.max())
        }
    }

    // Generate tick values
    val yTicks = remember(yMin, yMax) {
        val range = yMax - yMin
        val stride = when {
            range <= 12 -> 1f
            range <= 24 -> 2f
            else -> 3f
        }
        val ticks = mutableListOf<Float>()
        var current = ceil(yMin)
        while (current <= yMax) {
            ticks.add(current)
            current += stride
        }
        ticks
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(surfaceColor, MaterialTheme.shapes.small)
            .padding(8.dp)
    ) {
        if (title != null) {
            Text(
                text = title,
                style = MaterialTheme.typography.labelSmall,
                color = onSurfaceColor
            )
            Spacer(modifier = Modifier.height(4.dp))
        }

        Row(modifier = Modifier.fillMaxWidth()) {
            // Y-axis labels
            Column(
                modifier = Modifier
                    .width(32.dp)
                    .height(height.dp),
                verticalArrangement = Arrangement.SpaceBetween
            ) {
                yTicks.reversed().forEach { midi ->
                    Text(
                        text = CalibraMusic.midiToNoteLabel(midi),
                        style = MaterialTheme.typography.labelSmall,
                        fontSize = 8.sp,
                        color = onSurfaceColor
                    )
                }
            }

            // Graph canvas
            Canvas(
                modifier = Modifier
                    .weight(1f)
                    .height(height.dp)
                    .background(MaterialTheme.colorScheme.surface, MaterialTheme.shapes.small)
            ) {
                val canvasWidth = size.width
                val canvasHeight = size.height

                // Draw grid lines
                yTicks.forEach { midi ->
                    val y = mapY(midi, yMin, yMax, canvasHeight)
                    drawLine(
                        color = gridColor,
                        start = Offset(0f, y),
                        end = Offset(canvasWidth, y),
                        strokeWidth = 1f
                    )
                }

                if (dataPoints.isEmpty()) return@Canvas

                // Group by segment and draw each
                val segments = dataPoints.groupBy { it.segmentId }
                segments.forEach { (_, segmentPoints) ->
                    if (segmentPoints.size < 2) return@forEach

                    val seriesName = segmentPoints.first().series
                    val color = seriesColors[seriesName] ?: Color.Gray

                    val path = Path()
                    var started = false

                    segmentPoints.forEach { point ->
                        val x = mapX(point.time, timeMin, timeMax, canvasWidth)
                        val y = mapY(point.pitch, yMin, yMax, canvasHeight)

                        if (!started) {
                            path.moveTo(x, y)
                            started = true
                        } else {
                            path.lineTo(x, y)
                        }
                    }

                    drawPath(
                        path = path,
                        color = color,
                        style = Stroke(width = 2f)
                    )
                }
            }
        }

        // X-axis label
        Text(
            text = "Time (s)",
            style = MaterialTheme.typography.labelSmall,
            color = onSurfaceColor,
            modifier = Modifier.padding(start = 32.dp, top = 4.dp)
        )

        // Legend for multiple series
        if (seriesColors.size > 1) {
            Spacer(modifier = Modifier.height(4.dp))
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                seriesColors.forEach { (label, color) ->
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        modifier = Modifier.padding(start = 32.dp)
                    ) {
                        Canvas(modifier = Modifier.size(12.dp)) {
                            drawLine(
                                color = color,
                                start = Offset(0f, size.height / 2),
                                end = Offset(size.width, size.height / 2),
                                strokeWidth = 2f
                            )
                        }
                        Text(
                            text = label,
                            style = MaterialTheme.typography.labelSmall,
                            color = onSurfaceColor
                        )
                    }
                }
            }
        }
    }
}

private fun mapX(value: Float, min: Float, max: Float, canvasWidth: Float): Float {
    if (max == min) return canvasWidth / 2
    return ((value - min) / (max - min)) * canvasWidth
}

private fun mapY(value: Float, min: Float, max: Float, canvasHeight: Float): Float {
    if (max == min) return canvasHeight / 2
    return canvasHeight - ((value - min) / (max - min)) * canvasHeight
}

/**
 * Comparison graph for raw vs processed pitch contours.
 */
@Composable
fun PitchComparisonGraphView(
    rawPitches: List<Float>,
    processedPitches: List<Float>,
    times: List<Float>? = null,
    rawLabel: String = "Raw",
    processedLabel: String = "Processed",
    height: Int = 200
) {
    PitchGraphView(
        contours = listOf(
            ContourData(rawPitches, Color.Gray, rawLabel),
            ContourData(processedPitches, Color.Blue, processedLabel)
        ),
        times = times,
        title = "Raw vs Processed",
        height = height
    )
}

/**
 * Overload that accepts FloatArray and Dp height.
 */
@Composable
fun PitchGraphView(
    pitchesHz: FloatArray,
    times: FloatArray? = null,
    color: Color = Color.Blue,
    title: String? = null,
    series: String = "Pitch",
    height: Dp = 200.dp,
    inputIsMidi: Boolean = false
) {
    PitchGraphView(
        pitchesHz = pitchesHz.toList(),
        times = times?.toList(),
        color = color,
        title = title,
        series = series,
        height = height.value.toInt(),
        inputIsMidi = inputIsMidi
    )
}
